bpf/cpu/cpu.bpf.c

```diff
@@ -58,24 +54,217 @@ typedef __u64 stack_trace_type[MAX_STACK_DEPTH];
+// `stack_unwind_row`` to multiple buffers containing each field. That way we
+// would be able to not only have more entries, but we would increase
+// performance as more data will be able to fit in the CPU cache.
+//
+// This is particularly important for the program counter => map<pid, pcs> +
+// map<pid, other_data>. the second map can be split further if we decide to do
+// so.
+//
+// This is at the cost of code readability, so should only be done if
+// experiments confirm this theory.
+//
+// PERF(javierhonduco): Some of these types use a bigger type than we need, but
+// this makes prototyping easier as no padding should be added between fields.
+// Later on, we can make this more compact, which again, will allow us to pack
+// more items + make a better use of the CPU caches.
+typedef struct stack_unwind_row {
+  u64 pc;
+  u64 cfa_reg;
+  s64 cfa_offset;
+  s64 rbp_offset;
+} stack_unwind_row_t;
+
+// Unwinding table representation.
+typedef struct stack_unwind_table_t {
+  u64 table_len; // size of the table, as the max size is static.
+  stack_unwind_row_t rows[MAX_UNWIND_TABLE_SIZE];
+} stack_unwind_table_t;
+
+// Context for the binary search. We mutate it on every step and at the end
+// we read the resulting values.
+struct callback_ctx {
+  u64 pc;                      // the needle.
+  stack_unwind_table_t *table; // the haystack.
+  u32 found;
+  u32 left;  // current left index.
+  u32 right; // current right index.
+};
+
+// TODO(javierhonduco): Improve register list, this is just the two
+// registers we need for x86_64.
+enum registers {
+  X86_64_REGISTER_RBP = 6,
+  X86_64_REGISTER_RSP = 7,
+};
+
+// Statistics.
+//
+// We reached main.
+u32 UNWIND_SUCCESS = 1;
+// Partial stack was retrieved.
+u32 UNWIND_TRUNCATED = 2;
+// An (unhandled) dwarf expression was found.
+u32 UNWIND_UNSUPPORTED_EXPRESSION = 3;
+// Any other error, such as failed memory reads.
+// TODO(javierhonduco): split this error into subtypes.
+u32 UNWIND_CATCHALL_ERROR = 4;
+// Errors that should never happen.
+u32 UNWIND_SHOULD_NEVER_HAPPEN_ERROR = 5;
+// PC not in table (Kernel PC?).
+u32 UNWIND_PC_NOT_COVERED_ERROR = 6;
+// Keep track of total samples.
+u32 UNWIND_SAMPLES_COUNT = 7;
 /*================================ MAPS =====================================*/
 
-BPF_HASH(counts, stack_count_key_t, u64);
+BPF_HASH(counts, stack_count_key_t, u64, MAX_ENTRIES);
+
 BPF_STACK_TRACE(stack_traces, MAX_STACK_ADDRESSES);
+BPF_HASH(unwind_tables, pid_t, stack_unwind_table_t, MAX_PID_MAP_SIZE);
+
+struct {
+  __uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+  __uint(max_entries, 1);
+  __type(key, __u32);
+  __type(value, stack_count_key_t);
+} heap SEC(".maps");
+
+struct {
+  __uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+  __uint(max_entries, 10);
+  __type(key, __u32);
+  __type(value, __u32);
+} percpu_stats SEC(".maps");
 
 /*=========================== HELPER FUNCTIONS ==============================*/
 
+static void unwind_success() {
```

Comment:
I realise this is a lot of duplication...

Reply:
Yup, might be good to optimize this duplicate code. Doesn't have to be in this PR though.

Reply:
It seems like a good fit for a macro.